### DISCUSSION TITLE
Easily auto-create a Program Guide (EPG) for media centers from your AzuraCast!

### DISCUSSION THREAD

**Discussion by Moonbase59** [AzuraCast/AzuraCast#6677] [Show and tell]
## Ever wanted more distribution, easy access to your streams in a media center like Jellyfin, KODI, Plex or Emby? Plus a nice-looking Electronic Program Guide with an overview of your scheduled shows?

**Look no further.** This was something that nagged my for a long time, so I have written [`azuracast_xmltv`](https://github.com/Moonbase59/azuracast_xmltv).

It uses the well-respected XMLTV/IPTV standards to create "tuner" (M3U) and EPG (XML) files for all your stations. Automated. Just by pointing it to your AzuraCast instance.

It has a bunch of features (more if you give it an API key), and it’s Free and Open-Source Software, easy to use, and creates standards-compliant output.

But let screenshots speak (from my own installation):

![Live-TV – Mozilla Firefox_004](https://github.com/AzuraCast/AzuraCast/assets/3706922/feedb944-5273-4e48-bbe5-8e10e4ee271b)  
_azuracast_xmltv-generated EPG on [Jellyfin](https://jellyfin.org/)_

![screenshot00001](https://github.com/AzuraCast/AzuraCast/assets/3706922/def406dc-bd27-4463-86e7-2b7273416c56)  
_azuracast_xmltv-generated EPG on [KODI](https://kodi.tv/)_

![Moonbase_001](https://github.com/AzuraCast/AzuraCast/assets/3706922/22747661-d1f5-480b-99f4-b4b8e0430db4)  
_azuracast_xmltv-generated M3U on [Hypnotix](https://github.com/linuxmint/hypnotix), the Linux Mint IPTV player (no EPG yet)_

**Testers welcome.**

I’ve been talking to @BusterNeece, and we might evolve this even more together. But in the meantime, you have a tool that has been working beautifully here for several weeks now.

Read more about installation and features at https://github.com/Moonbase59/azuracast_xmltv.

**Comment by Moonbase59** [comment]
Added a README section on [how to install on a real AzuraCast server](https://github.com/Moonbase59/azuracast_xmltv/tree/master#install-on-a-real-azuracast-server-ubuntu-2204).

**Comment by Moonbase59** [comment]
@LordHelmchen666 could I interest you in some real-life testing?

Reasons I’m asking _you_ especially:
- you participate here and seem interested in evolving your station, even publish nice scripts
- you’re geographically near, we could even just pick up the phone
- you have a nice AzuraCast schedule
- I listen to your station sometimes (and actually use it as one of my "testbeds", just because I like having the station in my Jellyfin & KODI. `azuracast_xmltv`’s `User-Agent` is `azuracast_xmltv/x.y.z`, where `x.y.z` denote the version number. Just in case you want to check.)
- it might be useful to your station, getting more distribution & easy access (live broadcasts!)

All others: No offence, this is _not_ some personal judgment or devaluation of others. **Everyone** is invited for testing, actually **the more testers the better**. Your feedback is more than welcome. I’d very much like to know if station operators find this useful, how and why they use it!

**Reply by LordHelmchen666** [reply]
I will try it today.

**Reply by LordHelmchen666** [reply]
You can test https://stream.eisradio.de/xmltv/stream.eisradio.de.xml

**Reply by Moonbase59** [reply]
How cool is that! The second (I assume) station that uses /xmltv. I’ll update my xTeVe.

Want some suggestions?
- All the text including the `append_to_description` appears a little clumped. I’d use some newlines. (The `<desc>…</desc>` is actually the only one XML programme field where newlines are allowed.)
- Maybe the `append_to_description` isn’t needed after all, I just put some text in there to show what can be done.
- Maybe rename your 'MP3@192kbps' & 'AAC@32kbps' to something like 'Eisradio (MP3@192kbps)' & 'Eisradio (AAC@32kbps)' so they appear next to each other in other people’s EPG?

Works great, though, thanks for testing & using! Maybe we really get some "de-facto standard" together, many more stations involved using that, and get @BusterNeece interested in providing a schedule or playlist description field in the UI. ;-)

Btw, version 0.9.0 is up, fixing some things.

**Reply by Moonbase59** [reply]
Looks nice now:

```xml
  <programme start="20231022210000 +0200" stop="20231023000000 +0200" channel="eisradio.stream.eisradio.de">
    <title lang="en">Ballads (Musikwünsche erlaubt)</title>
    <sub-title lang="en">Deine Station, DEINE Musik!</sub-title>
    <desc lang="en">Playliste: Ballads

Machen Sie dieses Programm zu IHREM Programm, indem Sie einen Musikwunsch hinzufügen.
Gehen Sie zu https://stream.eisradio.de/public/eisradio, klicken Sie auf ›Lied wünschen‹ und wählen Sie Ihren Favoriten aus.

Program Copyright © 2023 Eisradio — Eisradio - Sport Radio und Community aus Regensburg (Oberpfalz/ Bayern). Community zum Sport in Regensburg.
Besuche uns https://www.eisradio.de/</desc>
    <credits/>
    <category lang="en">Music</category>
  </programme>
```

and validates ok:

```
$ wget -q https://stream.eisradio.de/xmltv/stream.eisradio.de.xml
$ tv_validate_file stream.eisradio.de.xml
Validated ok.
```

![screenshot00006](https://github.com/AzuraCast/AzuraCast/assets/3706922/e6d4f6b3-245b-45fb-8fb8-51fd07e5c910)

![screenshot00005](https://github.com/AzuraCast/AzuraCast/assets/3706922/1168a5d5-c74b-4fe8-b5fe-3449010fdf08)

![screenshot00007](https://github.com/AzuraCast/AzuraCast/assets/3706922/0600cdf7-19c9-4637-86d5-7c8ade74f345)

**Reply by Moonbase59** [reply]
Working late in the studio … a glance on the studio display … and suddenly realizing that KODI is still playing and you must have changed the stream names. Well done! :-) (Although the "Metal Night" is just your station ID and the "Song of the Bavarians" endlessly repeating.)

![Arbeitsfläche 1_035](https://github.com/AzuraCast/AzuraCast/assets/3706922/efad641e-ec26-40cb-a4d9-11929f160eda)  
_[StudioDisplay](https://github.com/Moonbase59/studiodisplay) shows my station on the left, KODI with Eisradio running on the right_

> »Ich liebe es, wenn ein Plan funktioniert.« (»I love it when a plan comes together.«)
> — _The A Team_

**Reply by LordHelmchen666** [reply]
Sometimes my playlist "Impressum" is going into an loop. It should play between 00:05 - 00:10. I  could not find the reason why. Maybe because it was too late playing at 00:28? Is working for days, weeks and suddenly looping. Should play three tracks "Station ID, Impressum, Bavaria Antheme" then switch to playlist "Metal Night" which runs most of the time very well.

**Comment by LordHelmchen666** [comment]
I have uptaded to 0.9.0. Everything works just fine.

**Reply by Moonbase59** [reply]
Love to hear that! (And 0.10.0 is next which adds gzip compression)
What clients are you using/testing it with?

**Comment by Moonbase59** [comment]
Hope we’ll see more AzuraCast stations with an EPG soon… 164 clones already.
Anyone else wanting to give feedback/suggestions?
